### DISCUSSION TITLE
yolov5/tutorials/multi_gpu_training/

### DISCUSSION THREAD

**Discussion by giscus** [ultralytics/ultralytics#10339] [Docs]
# yolov5/tutorials/multi_gpu_training/

Learn how to train datasets on single or multiple GPUs using YOLOv5. Includes setup, training modes and result profiling for efficient leveraging of multiple GPUs.

https://docs.ultralytics.com/yolov5/tutorials/multi_gpu_training/

**Comment by Zhaninh** [comment]
Could I use MacBook pro M1 and train multi-GPU?

**Reply by pderrenger** [reply]
Hey there! 🌟

Training on a MacBook Pro M1 using multiple GPUs isn't directly possible as the M1 chip is a single integrated unit with GPU cores included, and it doesn't support connecting external GPUs for distributed training like you might with traditional GPU setups. However, you can still harness the power of the M1's GPU for training, though you'd be limited to the capabilities of the M1 chip itself.

If you're looking to leverage multiple devices for training, you'd need to look into setups with external GPUs supported by platforms like Linux or Windows where you can utilize PyTorch's distributed training capabilities as outlined in our Multi-GPU Training guide:

```bash
python -m torch.distributed.run --nproc_per_node=2 train.py --batch 64 --data coco.yaml --weights yolov5s.pt --device 0,1
```

Remember, the M1's performance is quite robust for many tasks, so don't hesitate to run your models directly on it! If you find the need for more power or parallel GPU training, exploring a setup with traditional GPUs might be the next step. 🖥️💥

Happy training!

**Comment by kh-diab** [comment]
hello,
python -m torch.distributed.run --nproc_per_node 2 train.py --batch 64 --data coco.yaml --weights yolov5s.pt --device 0,1

here in this command what is the train.py what should be written there  is it model.train and if yes which parameters should be passed?

**Reply by glenn-jocher** [reply]
@kh-diab the `train.py` script in your command is the main training script for YOLOv5. It handles the training process, including loading the dataset, configuring the model, and running the training loop. You don't need to replace it with `model.train`. Instead, ensure you have the `train.py` script from the YOLOv5 repository in your working directory. The parameters you pass, such as `--batch`, `--data`, `--weights`, and `--device`, are correctly specified. For more details on training configurations, you can refer to the [Ultralytics YOLOv5 documentation](https://docs.ultralytics.com/yolov5/tutorials/multi_gpu_training/).

**Reply by kh-diab** [reply]
Thanks @glenn-jocher 
the documentation shows yolov5
I was using yolov8 
can I use v8 with multiple GPU?

**Reply by glenn-jocher** [reply]
@kh-diab yes, you can use YOLOv8 with multiple GPUs. The process is similar to YOLOv5. For YOLOv8, you can use the `torch.distributed.run` command to enable multi-GPU training. Here's an example command:

```bash
python -m torch.distributed.run --nproc_per_node 2 train.py --batch 64 --data coco.yaml --weights yolov8n.pt --device 0,1
```

Ensure you have the `train.py` script from the YOLOv8 repository in your working directory. This script will handle the training process across multiple GPUs. For more details, refer to the [Ultralytics YOLOv8 documentation](https://docs.ultralytics.com/modes/train/).

**Reply by kh-diab** [reply]
Hello @glenn-jocher 
thanks for the info, but I am still confused why should I clone the yolov8, 
I already installed everything using conda and pip such as ultralytics
how can I can use the multiple GPU in this case 
because when i try to pass parameter device =0,1,2,3 
it actually become slower

**Reply by kh-diab** [reply]
can you please send me the guide of how to use yolov8 to multiple GPU usage 
when I use device = 0,1,2,3 for example the results are not good and it become slower

**Reply by glenn-jocher** [reply]
To use YOLOv8 with multiple GPUs, you should use the `torch.distributed.run` command for efficient multi-GPU training. The `device=0,1,2,3` approach can be slower due to inefficient data parallelism. Instead, try the following command:

```bash
python -m torch.distributed.run --nproc_per_node 4 train.py --batch 64 --data coco.yaml --weights yolov8n.pt --device 0,1,2,3
```

This method ensures better utilization of GPU resources. If you encounter issues, please ensure you are using the latest version of the Ultralytics package. For more details, refer to the [Ultralytics YOLOv8 documentation](https://docs.ultralytics.com/modes/train/).

**Reply by kh-diab** [reply]
Thanks but again , I am not able to find the train.py where is it located, 
I installed yolo and ultralytics using conda and pip
can you guide me how to reach this file or from where is it based 
thanks in advance

**Reply by glenn-jocher** [reply]
The `train.py` script is part of the YOLOv8 repository and is used to handle the training process. Since you installed YOLO and Ultralytics via conda and pip, you don't need to clone the repository separately. Instead, you can use the `ultralytics` package directly in your Python environment. To train with multiple GPUs, you can use the following command:

```bash
python -m torch.distributed.run --nproc_per_node 4 -m ultralytics.yolo train --batch 64 --data coco.yaml --weights yolov8n.pt --device 0,1,2,3
```

This command leverages the `torch.distributed.run` utility for efficient multi-GPU training. If you encounter any issues, please ensure you are using the latest version of the Ultralytics package. For more details, refer to the [Ultralytics YOLOv8 documentation](https://docs.ultralytics.com/modes/train/).

**Comment by kh-diab** [comment]
if ultralytics is installed using conda not via git, what should i use instead of train.py

python -m torch.distributed.run --nproc_per_node 2 train.py --batch 5  --data /home/kdiab/Documents/OST/workspace/dataset/coding/data_group/data_group_A2.yaml --weights yolov8x.pt --device 1

**Reply by glenn-jocher** [reply]
@kh-diab if you have installed Ultralytics YOLO using conda, you can replace `train.py` with the `yolo` command for training. Your command would look like this:

```bash
python -m torch.distributed.run --nproc_per_node 2 yolo train --batch 5 --data /home/kdiab/Documents/OST/workspace/dataset/coding/data_group/data_group_A2.yaml --weights yolov8x.pt --device 1
```

This command leverages the `yolo` CLI, which is designed to streamline various tasks including training, validation, and prediction. If you encounter any issues, please ensure you are using the latest version of the Ultralytics package. For further details, you can refer to the [Ultralytics documentation](https://docs.ultralytics.com/modes/train/).

**Reply by kh-diab** [reply]
Thanks for your reply, however after checking the documentation, 
it is not so clear for me
there was no mention how to use the other parameters 
such as amp or cache

also tried the command you shared lately and it give me this error.



(yolov8) kdiab@robi:~/Documents/OST/workspace/dataset/coding$ python -m torch.distributed.run --nproc_per_node 2 yolo train --batch 5 --data /home/kdiab/Documents/OST/workspace/dataset/coding/data_group/data_group_C_Segment_640_yolo_v2.yaml --weights yolov8x-seg.pt --device 1,2
WARNING:__main__:
*****************************************
Setting OMP_NUM_THREADS environment variable for each process to be 1 in default, to avoid your system being overloaded, please further tune the variable for optimal performance in your application as needed. 
*****************************************
/home/kdiab/.conda/envs/yolov8/bin/python: can't open file '/home/kdiab/Documents/OST/workspace/dataset/coding/yolo': [Errno 2] No such file or directory
/home/kdiab/.conda/envs/yolov8/bin/python: can't open file '/home/kdiab/Documents/OST/workspace/dataset/coding/yolo': [Errno 2] No such file or directory


though I have changed to the yolo.pt file before running the script

**Reply by glenn-jocher** [reply]
It looks like the error you're encountering is due to the `yolo` command not being found. This can happen if the `yolo` CLI is not properly installed or not in your PATH. Ensure you have installed the Ultralytics package correctly via conda. You can try running `yolo` directly in your terminal to check if it's recognized. For additional parameters like `amp` or `cache`, you can include them directly in your command. If the issue persists, please verify the installation and try again with the latest package version. For more details, refer to the [Ultralytics documentation](https://docs.ultralytics.com/modes/train/).

**Reply by kh-diab** [reply]
hello i tried this

yolo train batch=20 data=/data_group_A2.yaml  model=yolov8x.pt device=0,1,3,4
but it throws exception


```
RuntimeError: CUDA error: CUDA-capable device(s) is/are busy or unavailable
CUDA kernel errors might be asynchronously reported at some other API call, so the stacktrace below might be incorrect.
For debugging consider passing CUDA_LAUNCH_BLOCKING=1.
Compile with `TORCH_USE_CUDA_DSA` to enable device-side assertions.

Transferred 589/595 items from pretrained weights
Freezing layer 'model.22.dfl.conv.weight'
AMP: running Automatic Mixed Precision (AMP) checks with YOLOv8n...
W0822 16:12:30.269268 140669819975488 torch/distributed/elastic/multiprocessing/api.py:851] Sending process 3737927 closing signal SIGTERM
W0822 16:12:30.269666 140669819975488 torch/distributed/elastic/multiprocessing/api.py:851] Sending process 3737928 closing signal SIGTERM
W0822 16:12:30.269770 140669819975488 torch/distributed/elastic/multiprocessing/api.py:851] Sending process 3737929 closing signal SIGTERM
E0822 16:12:30.897917 140669819975488 torch/distributed/elastic/multiprocessing/api.py:826] failed (exitcode: 1) local_rank: 3 (pid: 3737930) of binary: /usr/bin/python3
Traceback (most recent call last):
  File "/usr/lib/python3.8/runpy.py", line 194, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/usr/lib/python3.8/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/home/kdiab/.local/lib/python3.8/site-packages/torch/distributed/run.py", line 883, in <module>
    main()
  File "/home/kdiab/.local/lib/python3.8/site-packages/torch/distributed/elastic/multiprocessing/errors/__init__.py", line 347, in wrapper
    return f(*args, **kwargs)
  File "/home/kdiab/.local/lib/python3.8/site-packages/torch/distributed/run.py", line 879, in main
    run(args)
  File "/home/kdiab/.local/lib/python3.8/site-packages/torch/distributed/run.py", line 870, in run
    elastic_launch(
  File "/home/kdiab/.local/lib/python3.8/site-packages/torch/distributed/launcher/api.py", line 132, in __call__
    return launch_agent(self._config, self._entrypoint, list(args))
  File "/home/kdiab/.local/lib/python3.8/site-packages/torch/distributed/launcher/api.py", line 263, in launch_agent
    raise ChildFailedError(
torch.distributed.elastic.multiprocessing.errors.ChildFailedError: 
============================================================
/home/kdiab/.config/Ultralytics/DDP/_temp_kmy48de8139867604963136.py FAILED
------------------------------------------------------------
Failures:
  <NO_OTHER_FAILURES>
------------------------------------------------------------
Root Cause (first observed failure):
[0]:
  time      : 2024-08-22_16:12:30
  host      : robi
  rank      : 3 (local_rank: 3)
  exitcode  : 1 (pid: 3737930)
  error_file: <N/A>
  traceback : To enable traceback see: https://pytorch.org/docs/stable/elastic/errors.html
============================================================
Traceback (most recent call last):
  File "/home/kdiab/.local/bin/yolo", line 8, in <module>
    sys.exit(entrypoint())
  File "/home/kdiab/.local/lib/python3.8/site-packages/ultralytics/cfg/__init__.py", line 708, in entrypoint
    getattr(model, mode)(**overrides)  # default args from model
  File "/home/kdiab/.local/lib/python3.8/site-packages/ultralytics/engine/model.py", line 650, in train
    self.trainer.train()
  File "/home/kdiab/.local/lib/python3.8/site-packages/ultralytics/engine/trainer.py", line 199, in train
    raise e
  File "/home/kdiab/.local/lib/python3.8/site-packages/ultralytics/engine/trainer.py", line 197, in train
    subprocess.run(cmd, check=True)
  File "/usr/lib/python3.8/subprocess.py", line 516, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['/usr/bin/python3', '-m', 'torch.distributed.run', '--nproc_per_node', '4', '--master_port', '35103', '/home/kdiab/.config/Ultralytics/DDP/_temp_kmy48de8139867604963136.py']' returned non-zero exit status 1.
```

**Reply by glenn-jocher** [reply]
It seems you're encountering a CUDA error related to device availability. Ensure all specified GPUs are free and properly configured. You might also try setting `CUDA_LAUNCH_BLOCKING=1` for more detailed error messages. For further assistance, please refer to the [multi-GPU training guide](https://docs.ultralytics.com/yolov5/tutorials/multi_gpu_training/).

**Comment by sivikt** [comment]
Hey all,

is it ok to run Multi-GPU (1 host) training not from cli with `torch.distributed.run --nproc_per_node` but simply as `model.train(device=[1,2,3,4])`? 

I run as `model.train(device=[1,2,3,4])` and having several callback functions `on_pretrain_routine_end, on_fit_epoch_end, on_train_epoch_end` can not make them work: mlflow does not write anything into experiment.

**Reply by sivikt** [reply]
mlflow is setup beforehand, before running training as `MlflowClient()` (hosted in Databricks)

**Reply by glenn-jocher** [reply]
@sivikt using `model.train(device=[1,2,3,4])` for multi-GPU training is possible, but for optimal performance and callback functionality, it's recommended to use `torch.distributed.run`. This ensures proper synchronization and logging across GPUs. If you're experiencing issues with callbacks like `mlflow`, it might be due to the way processes are managed in non-distributed settings. For more details, you can refer to the [Ultralytics documentation](https://docs.ultralytics.com/modes/train/).

**Comment by RobertKoval** [comment]
Can I use multiple Mac mini for distributed training?

**Reply by glenn-jocher** [reply]
@RobertKoval yes, you can use multiple Mac minis for distributed training with YOLOv5 by setting up a multi-node environment using PyTorch's DistributedDataParallel. Ensure all machines have the same setup and can communicate with each other. For detailed instructions, refer to the multi-GPU training guide in the documentation.
